nodes/move_to_pregrasp.py

```diff
@@ -0,0 +1,953 @@
+                        return action_error_callback(
+                            f"Failed to solve IK {ik_solution}",
+                            MoveToPregrasp.Result.STATUS_GOAL_NOT_REACHABLE,
+                        )
+                    break
+            if not reachable:
+                break
+
+            # Move the robot
+            if len(motion_executors) == 0:
+                joints_for_velocity_control = []
+                joint_position_overrides = {}
+                joints_for_position_control = {}
+                velocity_overrides = {}
+                cartesian_mask = None
+                if MoveToPregraspState.TERMINAL in concurrent_states:  # TERMINAL
+                    return action_success_callback("Goal succeeded")
+                if MoveToPregraspState.STOW_ARM_LENGTH_FULL in concurrent_states:
+                    joints_for_position_control.update(
+                        get_stow_configuration([Joint.ARM_L0])
+                    )
+                if MoveToPregraspState.STOW_ARM_LENGTH_PARTIAL in concurrent_states:
+                    joints_for_position_control.update(
+                        get_stow_configuration([Joint.ARM_L0], partial=True)
+                    )
+                if MoveToPregraspState.STOW_ARM_LIFT in concurrent_states:
+                    joints_for_position_control.update(
+                        get_stow_configuration([Joint.ARM_LIFT])
+                    )
+                if MoveToPregraspState.STOW_WRIST in concurrent_states:
+                    joints_for_position_control.update(
+                        get_stow_configuration(
+                            Joint.get_wrist_joints() + [Joint.GRIPPER_LEFT]
+                        )
+                    )
+                if MoveToPregraspState.ROTATE_BASE in concurrent_states:
+                    # TODO: This sometimes ends up a few degrees off. The object is
+                    # still in the gripper camera view, so users can resolve this,
+                    # but we should look into why (odom -> base_link TF delay?
+                    # issue with termination thresholds? robot's center of rotation? etc.)
+                    joints_for_velocity_control += [Joint.BASE_ROTATION]
+                    joint_position_overrides.update(
+                        {
+                            joint: position
+                            for joint, position in ik_solution.items()
+                            if joint != Joint.BASE_ROTATION
+                        }
+                    )
+                    joint_position_overrides.update(
+                        get_pregrasp_wrist_configuration(horizontal_grasp)
+                    )
+                    # Only care about x and yaw in base frame. y and z are adjusted by
+                    # arm length/lift
+                    cartesian_mask = np.array([True, False, False, False, False, True])
+                if MoveToPregraspState.HEAD_PAN in concurrent_states:
```

Comment:
why is current states multiple things? i guess you can execute multiple steps at once in this model?

Reply:
Yes, this is specifically because of a feature Blaine requested (that is part of this PR) to rotate the head pan to counteract base rotation. So those states are executed at the same time.

More generally, if in the future we want to make this less step-wise and move multiple joints together, we can use this "combining states" feature.